orderer/common/follower/follower_chain.go

```diff
@@ -286,146 +255,277 @@ func (c *Chain) run() {
+			latestNetworkHeight, retryInterval)
+		select {
+		case <-c.stopChan:
+			c.logger.Debug("Received a stop signal")
+			return ErrChainStopped
+		case <-c.timeAfter(retryInterval):
+			// Exponential back-off on successive errors w/o progress.
+			c.increaseRetryInterval(&retryInterval, c.options.PullRetryMaxInterval)
+		}
+	}
+
+	return nil
+}
+
+// pullUntilTarget is given a target-height and exits without an error when it reaches that target.
+// It may return with an error before the target, always returning the number of blocks pulled.
+func (c *Chain) pullUntilTarget(targetHeight uint64, inspectConfig bool) (uint64, error) {
+	firstBlockToPull := c.ledgerResources.Height()
+	if firstBlockToPull >= targetHeight {
+		c.logger.Debugf("Target height (%d) is <= to our ledger height (%d), skipping pulling", targetHeight, firstBlockToPull)
+		return 0, nil
+	}
 
 	var actualPrevHash []byte
 	// Initialize the actual previous hash
 	if firstBlockToPull > 0 {
 		prevBlock := c.ledgerResources.Block(firstBlockToPull - 1)
 		if prevBlock == nil {
-			return errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
+			return 0, errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
 		}
 		actualPrevHash = protoutil.BlockHeaderHash(prevBlock.Header)
 	}
 
-	// Pull the rest of the blocks
+	// Pull until the latest height
 	for seq := firstBlockToPull; seq < targetHeight; seq++ {
+		n := seq - firstBlockToPull
 		select {
 		case <-c.stopChan:
-			c.logger.Warnf("Stopped before pulling all the blocks: pulled %d blocks from the range %d until %d",
-				seq-firstBlockToPull, firstBlockToPull, targetHeight-1)
-			return errors.New("stopped before pulling all the blocks")
+			c.logger.Debug("Received a stop signal")
+			return n, ErrChainStopped
 		default:
-			nextBlock := puller.PullBlock(seq)
+			nextBlock := c.blockPuller.PullBlock(seq)
 			if nextBlock == nil {
-				return errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
+				return n, errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
 			}
 			reportedPrevHash := nextBlock.Header.PreviousHash
 			if (nextBlock.Header.Number > 0) && !bytes.Equal(reportedPrevHash, actualPrevHash) {
-				return errors.Errorf("block header mismatch on sequence %d, expected %x, got %x",
+				return n, errors.Errorf("block header mismatch on sequence %d, expected %x, got %x",
 					nextBlock.Header.Number, actualPrevHash, reportedPrevHash)
 			}
 			actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
-			if err = c.ledgerResources.Append(nextBlock); err != nil {
-				return errors.Wrapf(err, "failed to append block %d to the ledger", nextBlock.Header.Number)
+			if err := c.ledgerResources.Append(nextBlock); err != nil {
+				return n, errors.Wrapf(err, "failed to append block %d to the ledger", nextBlock.Header.Number)
+			}
+
+			if inspectConfig && protoutil.IsConfigBlock(nextBlock) {
```

Comment:
To confirm my understanding of why this is safe:

The block puller is created only using the orderer endpoints, as defined in the channel config.  Internally, the block puller through, magic or other means, uses the correct channel config for performing block signature validation.

Here, we are performing a secondary check, that, if the block we just retrieved changes the channel config, we are using this new channel config in order to generate a new block puller with more recent updates.

In the pullFromJoin path, we set `inspectConfig` to false, we already have 'more up to date' endpoints, but in the `pullAfterJoin` path, we set `inspectConfig` to true, to ensure that we re-construct the block puller with the most up to date endpoints.

Assuming that's all correct, what I'd like some clarification on, is if this is strictly necessary, and if this has any performance implications.  If we just got a config block, from the set of endpoints we know about, then it would imply to me, that the set of endpoints we currently know about is adequate / working.  In particular, I assume that the block puller, once connected, is pulling from a single endpoint, and because we just received a block, if we were to continue without recreating the block puller, nothing would break.  Is this because the block puller internally has error handling that will seek out and fail over to another node if this one fails?

Reply:
The current onboarding or Raft replication doesn't re-create the block puller upon [ingestion of a config block](https://github.com/hyperledger/fabric/blob/master/orderer/common/cluster/util.go#L535-L561). This is not necessary because the block puller [knows to update its internal bundle and PolicyManager](https://github.com/hyperledger/fabric/blob/87491e68c289b18f871527113c356afdaef36484/orderer/common/cluster/util.go#L617-L635) in-flight and autonomously. 
When you create a new block puller, it takes the bundle and PolicyManager that is the latest persisted on disk.
Then, as it eats config blocks, [it replaces this bundle and policy manager](https://github.com/hyperledger/fabric/blob/87491e68c289b18f871527113c356afdaef36484/orderer/common/cluster/util.go#L593-L606) reference.

Reply:
Your understanding is mostly correct. Let me shed some more light and answer some of the questions.

Here is the overall desired behavior:
1. When pulling until the join block, the puller should take the most updated endpoints (from the join block), but use a block signature verification policy from the config block just prior to the block is it trying to verify. We cannot verify block `m` with the channel config from join-block `n` if `m<n` (it might work for degenerate cases, like our current integration tests, but it's wrong). Let's call this component the `blockVerifier`. This means that the `blockVerifier` need to be built from the genesis block up, until we reach the join block, but the endpoints are just from the join-block.
2. When pulling past the join block, we have to update the `blockVerifier` with every new channel config, and as we do so, we might as well do the same for the endpoints.

Your comment about the "magic" by which the block puller updates the channel config for signature verification sent me checking (again) on how this is currently implemented in components using the block puller in fabric. 
The conclusion is, that this does not happen "internally" to the block puller but rather on an upper layer.

- The etcdraft.Chain, in the [catchUp code](https://github.com/hyperledger/fabric/blob/9c705811e9d2dbda6a7876d0065d65a369a62ca5/orderer/consensus/etcdraft/chain.go#L879), provides the "support" to the block puller. This has the `ledgerResources` which has the `mutableConfig`, which exposes the `Update(bundle)`. It then calls the block-writer `WriteConfigBlock` on every new config which internally updates the channel config after commit. Note that the block puller is built from an opaque function pointer, which is in fact a closure defined in the registrar.
- The `onboarding` package that implements the `InactiveChainReplicator` does it differently - it maintains a separate `blockVerifier` for every channel, and updates it after every `Append`. (see the [verifierLoader](https://github.com/hyperledger/fabric/blob/9c705811e9d2dbda6a7876d0065d65a369a62ca5/orderer/common/onboarding/onboarding.go#L358)  the [VerifierFactory](https://github.com/hyperledger/fabric/blob/9c705811e9d2dbda6a7876d0065d65a369a62ca5/orderer/common/cluster/util.go#L493) and the [BlockVerifierAssembler](https://github.com/hyperledger/fabric/blob/9c705811e9d2dbda6a7876d0065d65a369a62ca5/orderer/common/cluster/util.go#L587)).

The the main difference is that the `etcdraft.Chain` has a `support`  with the `ledgerResources` & `blockWriter` in it, whereas the `onboarding.Inactive*` does not. 

My original plan was to use different block pullers, constructed by the Registrar, depending on whether we construct the follower pre- or post-join. Hence the callback to recreate the follower to a new follower in the previous version. Specifically, after join, we can use the ledgerSources and its ability to `Update` to keep track of channel config. For pulling before the join, my plan was to use the block puller more like the `onboarding.Inactive*`, since we build the ledger resources from the join block, but pull config blocks from the past. 

Currently, all that wiring is not entirely in place. The `blockVerifier` given to the puller is based on the join block and is not updated. This will work is tests when channel config does not significantly change, but break otherwise. My plan is to put all that in place in the next [task / PR](https://jira.hyperledger.org/browse/FAB-18086), which I have already started. 

With this in mind, the way I would like to progress is the following. I propose to express the rest of the comments on this PR in a Jira (say "Improving follower flow control"), and address the issues around [wiring](https://jira.hyperledger.org/browse/FAB-18086) first. After those issues are nailed down, and the puller interface and usage patterns had stabilized, it will be much easier to implement the suggestions for flow control consolidation brought here.

Reply:
> Attending to the "external" functionality first would not only simplify the internal refactoring later, it would also allow the rest of the team to execute in parallel some of the tasks that are blocked by this one, for example integration testing. Refactoring internals with tests in place is also a good idea.

My feeling is still that we could drastically simplify this now.  But, I don't want to block progress longer than I already have.  I'll approve so that we can get things moving along, but I would like to make sure we revisit this in the future.